xla/service/gpu/runtime/nccl_collective_permute_thunk.h

```diff
@@ -46,26 +46,34 @@ class NcclCollectivePermuteStartThunk : public NcclCollectiveThunk {
   NcclCollectivePermuteStartThunk(ThunkInfo thunk_info, NcclApi* nccl_api,
                                   const HloCollectivePermuteInstruction* instr,
                                   int64_t replica_count,
-                                  int64_t partition_count,
-                                  const Buffer& buffer);
+                                  int64_t partition_count, const Buffer& buffer,
+                                  bool p2p_memcpy_enabled);
+  absl::Status Initialize(const InitializeParams& params) override;
+
+  absl::Status Cleanup(const CleanupParams& params) override;
 
   static const char* GetHloOpName() { return "collective-permute-start"; }
 
  protected:
   const NcclCollectiveConfig& config() const override { return config_.config; }
   absl::Status RunNcclCollective(const ExecuteParams& params,
                                  se::Stream& stream,
-                                 NcclApi::NcclCommHandle comm) override;
+                                 NcclApi::NcclCommHandle comm,
+                                 bool is_local) override;
 
  private:
   const NcclP2PConfig config_;
   const Buffer buffer_;
+  std::unordered_map<int64_t, uint64_t> send_value_map_;
+  std::unordered_map<int64_t, uint64_t> recv_value_map_;
```

Comment:
Also I'm not sure if you need cleanup stage at all, you can maintain `absl::node_hash_map<se::StreamExecutor*, void*>` map that you need to deregister in destructor. NCCL clique locking guarantees that you can't execute this thunk concurrently on the same stream executor. Once you see a new instance of stream executor, you register a pointer with it, and issue send/recv-ptr using that pointer.

Reply:
Yea, AsyncValue + map got rid of all the nccl calls so we dont need to register any host pointer anymore, therefore no need for cleanup or storing stream executor.